### ISSUE TITLE
Investigate why `sct_deepseg_sc` and `sct_get_centerline` produce different results

### ISSUE THREAD

**Issue by joshuacwnewton** [spinalcordtoolbox/spinalcordtoolbox#3993]
> _**NB:** This issue corresponds to one of my action items from the SCT meeting minutes. See the [2022-11-17](https://docs.google.com/document/d/1ItApJQfajO2lRzOU2yenWbeRg6alfsdut3J4AVVdo78/edit#heading=h.w2w1y4v0zcpa) minutes for more details._

Many CLI scripts in SCT use spinal cord centerline images in their processing. These scripts often determine the centerline by using the [`get_centerline()`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/a15070f635264ad05aae41db83ce31160600e930/spinalcordtoolbox/centerline/core.py#L85-L244) function. This function provides a number of methods (e.g. `optic`, `polyfit`, `bspline`, `nurbs`) for determining the centerline.

However, different CLI scripts use different default values:

* `sct_get_centerline`: Defaults to [`optic`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/a15070f635264ad05aae41db83ce31160600e930/spinalcordtoolbox/scripts/sct_get_centerline.py#L53).
* `sct_deepseg_sc`: Defaults to [`svm`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/a15070f635264ad05aae41db83ce31160600e930/spinalcordtoolbox/scripts/sct_deepseg_sc.py#L59). 
    * Confusingly, though, `svm` [actually means `optic`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/a15070f635264ad05aae41db83ce31160600e930/spinalcordtoolbox/deepseg_/sc.py#L52-L59). So, are these actually the same?

Furthermore, it was reported in the [2022-11-17](https://docs.google.com/document/d/1ItApJQfajO2lRzOU2yenWbeRg6alfsdut3J4AVVdo78/edit#heading=h.w2w1y4v0zcpa) meeting minutes that:

> “sct_deepseg_sc -file-centerline” + “sct_get_centerline” = better results?

In other words, `sct_deepseg_sc` and `sct_get_centerline` seem to be doing something different in how they call the `'optic'` method. 

This issue is for investigating these differences more closely, and determining how we can unify the performance between scripts.

**Comment by joshuacwnewton** [comment]
> “sct_deepseg_sc -file-centerline” + “sct_get_centerline” = better results?

Focusing on issue 2, the original report was in the context of the dog template project (i.e. the input data was challenging/unique).

For now, though, I am just going to run a very basic test using our (healthy, human) `sct_example_data` dataset:

```bash
# Prepare data
sct_download_data -d sct_example_data
cd $SCT_DIR/data/sct_example_data/t2
# Method 1: sct_deepseg_sc with defaults
sct_deepseg_sc -i t2.nii.gz -c t2 -ofolder 1_defaults
# Method 2: Use sct_get_centerline
sct_get_centerline -i t2.nii.gz -c t2
sct_deepseg_sc -i t2.nii.gz -c t2 -file_centerline t2_centerline.nii.gz -ofolder 2_centerline-file
# Subtract to compare:

```
Subtracting the two output `t2_seg.nii.gz` files, I can confirm that different results are produced (R/Y voxels are -1/+1):

<details>

![diff_screenshot](https://user-images.githubusercontent.com/16181459/212137286-6d956c5e-6fc8-4297-ae47-d338e990add7.png)

</details>

I'll see if I can figure out why next. :)

**Comment by joshuacwnewton** [comment]
I set a breakpoint in `get_centerline`, then re-ran both the `sct_deepseg_sc` call and the `sct_get_centerline` call to compare the state of the variables inside the function.

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/a15070f635264ad05aae41db83ce31160600e930/spinalcordtoolbox/centerline/core.py#L89-L92

I'm noticing a few differences here: 

1. In the `ParamCenterline` variable, the value of `smooth` is `20` for `sct_deepseg_sc`, but it's `30` for `sct_get_centerline`.
    * However, `smooth` is only relevant for the `bspline` and `linear` fitseg methods. (We're using `optic` by default.)
2. The `im_seg` variable is quite different between the two scripts:
    * `sct_deepseg_sc` [resamples](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/a15070f635264ad05aae41db83ce31160600e930/spinalcordtoolbox/deepseg_/sc.py#L434-L436) the input image before passing it to `get_centerline()`. 
        * `dim = (102, 512, 320, 1, 0.50196105, 0.5, 0.8, 1)`
    * `sct_get_centerline` passes the input image directly to `get_centerline()`, no resampling. 
        * `dim = (64, 320, 320, 1, 0.8000004, 0.8, 0.8, 1)`

So, if I modify `sct_deepseg_sc` to pass the exact original input image directly to `get_centerline()`, then resample the centerline after the fact (just like what is done for `-file-centerline`), I can produce 100% identical results between the two methods.

**Comment by joshuacwnewton** [comment]
In our [2023-01-12 meeting](https://docs.google.com/document/d/1ItApJQfajO2lRzOU2yenWbeRg6alfsdut3J4AVVdo78/edit#heading=h.2xefygcyqgee), @benjamindeleener suggested (paraphrasing) that resampling the input image has benefits for low resolution images, but negligible downsides for high-resolution images.

So, if we wanted to unify these scripts to produce identical results, we might consider make resampling inherent to `get_centerline()`, that way the input resolution is consistent regardless of how `get_centerline()` is called.

(One other benefit of this is that it would leave the results of `sct_deepseg_sc` unchanged, which I would guess is more popular than the `sct_get_centerline` script.)

However, @jcohenadad suggested that this issue is relatively low-priority compared to other issues/efforts; this is now a well-understood issue, and we can guide our students around the issue should they run into it in the future.
